### PR TITLE
executor: fix incorrect result when hash agg spill is triggered

### LINKED ISSUE #55290: hash agg parallel spill get wrong result
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
for a simple query like 
```
select count(*), id from t group by id
```
if parallel spill is triggered, it will return wrong result.

### PR DESCRIPTION
**Pull request by xzhangxian1008**
### What problem does this PR solve?

Issue Number: close #55290

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Fix the issue that getting incorrect result when hash agg spill is triggered
```

### CODE REVIEW

File: pkg/executor/aggfuncs/func_max_min.go [hunk 1]
```diff
@@ -1287,6 +1287,26 @@ func (e *maxMin4Time) MergePartialResult(_ AggFuncUpdateContext, src, dst Partia
 	return 0, nil
 }
 
+func (e *maxMin4Time) SerializePartialResult(partialResult PartialResult, chk *chunk.Chunk, spillHelper *SerializeHelper) {
```
Comment (XuHuaiyu) [inline pkg/executor/aggfuncs/func_max_min.go:1290]: Can we add test cases for more filedtypes ?

**Review by XuHuaiyu — APPROVED** [review]

File: pkg/executor/aggregate/agg_spill.go [hunk 2]
```diff
@@ -239,23 +243,23 @@ type processRowContext struct {
...
+	aggRestoreFuncNum             int
 	restoreadData          *aggfuncs.AggPartialResultMapper
 	partialResultsRestored [][]aggfuncs.PartialResult
 	bInMap                 *int
 }
 
 func (p *parallelHashAggSpillHelper) restoreFromOneSpillFile(ctx sessionctx.Context, restoreadData *aggfuncs.AggPartialResultMapper, diskIO *chunk.DataInDiskByChunks, bInMap *int) (totalMemDelta int64, totalExpandMem int64, err error) {
 	chunkNum := diskIO.NumChunks()
-	aggFuncNum := len(p.aggFuncsForRestoring)
+	aggRestoreFuncNum := len(p.aggFuncsForRestoring)
```
Comment (windtalker) [inline pkg/executor/aggregate/agg_spill.go:254]: when will len(p.aggFuncsForRestoring) != len(p.finalWorkerAggFuncs) ?

Comment (xzhangxian1008) [reply pkg/executor/aggregate/agg_spill.go:259]: > when will len(p.aggFuncsForRestoring) != len(p.finalWorkerAggFuncs) ?

fixed

File: pkg/executor/aggregate/agg_spill.go [hunk 3]
```diff
@@ -239,23 +248,23 @@ type processRowContext struct {
 	chunk                  *chunk.Chunk
 	rowPos                 int
 	keyColPos              int
-	aggFuncNum             int
+	aggRestoreFuncNum      int
```
Comment (windtalker) [inline pkg/executor/aggregate/agg_spill.go:251]: I think there is no need to rename this variable?

File: pkg/executor/aggregate/agg_spill.go [hunk 4]
```diff
@@ -293,8 +302,8 @@ func (p *parallelHashAggSpillHelper) processRow(context *processRowContext) (tot
 	if ok {
 		exprCtx := context.ctx.GetExprCtx()
 		// The key has appeared before, merge results.
-		for aggPos := 0; aggPos < context.aggFuncNum; aggPos++ {
-			memDelta, err := p.aggFuncsForRestoring[aggPos].MergePartialResult(exprCtx.GetEvalCtx(), context.partialResultsRestored[aggPos][context.rowPos], prs[aggPos])
+		for aggPos := 0; aggPos < len(p.finalWorkerAggFuncs); aggPos++ {
```
Comment (windtalker) [inline pkg/executor/aggregate/agg_spill.go:305]: use `aggPos < context.aggFuncNum` is enough?

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**